libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt

```diff
@@ -0,0 +1,113 @@
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.navigation.base.route.RouteAlternativesOptions
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.routeoptions.RouteOptionsUpdater
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.utils.internal.MapboxTimer
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArraySet
+
+internal class RouteAlternativesController(
+    private val options: RouteAlternativesOptions,
+    private val navigator: MapboxNativeNavigator,
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsUpdater: RouteOptionsUpdater,
+    private val logger: Logger
+) {
+    private val jobControl = ThreadController.getMainScopeAndRootJob()
+
+    private val mapboxTimer = MapboxTimer()
+    private val observers = CopyOnWriteArraySet<RouteAlternativesObserver>()
+    private var currentRequestId: Long? = null
+
+    fun register(routeAlternativesObserver: RouteAlternativesObserver) {
+        val needsToStartTimer = observers.isEmpty()
+        observers.add(routeAlternativesObserver)
+        if (needsToStartTimer) {
+            mapboxTimer.restartAfterMillis = options.intervalMillis
+            mapboxTimer.startTimer {
+                requestRouteAlternatives()
+            }
+        }
+    }
+
+    fun unregister(routeAlternativesObserver: RouteAlternativesObserver) {
+        observers.remove(routeAlternativesObserver)
+        if (observers.isEmpty()) {
+            stop()
+        }
+    }
+
+    fun unregisterAll() {
+        observers.clear()
+        stop()
+    }
+
+    private fun stop() {
+        mapboxTimer.stopJobs()
+        jobControl.job.cancelChildren()
+        currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+    }
+
+    private fun requestRouteAlternatives() {
+        if (directionsSession.routes.isEmpty()) {
+            return
+        }
+
+        val routeOptionsResult = routeOptionsUpdater.update(
+            directionsSession.getPrimaryRouteOptions(),
+            tripSession.getRouteProgress(),
+            tripSession.getEnhancedLocation()
+        )
+
+        when (routeOptionsResult) {
+            is RouteOptionsUpdater.RouteOptionsResult.Success -> {
+                currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+                currentRequestId = directionsSession.requestRoutes(
+                    routeOptionsResult.routeOptions,
+                    routesRequestCallback
+                )
+            }
+            is RouteOptionsUpdater.RouteOptionsResult.Error -> Unit
```

Comment:
Log an error.

Reply:
These errors are logged by [MapboxRouteOptionsUpdater](https://github.com/mapbox/mapbox-navigation-android/blob/bc1dce0cb8cba9ea8c8efc40bd8d65d5350a9d23/libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsUpdater.kt#L31-L37)

But yeah, doesn't hurt to have another. added